highway_env/envs/u_turn_env.py

```diff
@@ -0,0 +1,216 @@
+    """
+    U-Turn risk analysis task: the agent overtakes vechiles that are blocking the
+    traffic. High speed overtaking must be balanced with ensuring safety.
+    """
+
+    """Penalization received for vechile collision."""
+    COLLISION_REWARD: float = -1.0
+    """Reward received for maintaining left most lane."""
+    LEFT_LANE_REWARD: float = 0.1
+    """Reward received for maintaining cruzing speed."""
+    HIGH_SPEED_REWARD: float = 0.4
+
+    @classmethod
+    def default_config(cls) -> dict:
+        config = super().default_config()
+        config.update({
+            "observation": {
+                "type": "TimeToCollision",
+                "horizon": 16
+            },
+            "action": {
+                "type": "DiscreteMetaAction",
+            },
+            "screen_height": 789,  
+            "screen_height": 289,
+            "duration": 8,
+            "reward_speed_range": [16, 24],
+            "offroad_terminal": False
+        })
+        return config
+
+    def _reward(self, action: int) -> float:
+        """
+        The vehicle is rewarded for driving with high speed and collision avoidance.
+        :param action: the action performed
+        :return: the reward of the state-action transition
+        """
+        neighbours = self.road.network.all_side_lanes(self.vehicle.lane_index)
+        lane = self.vehicle.lane_index[2]
+        scaled_speed = utils.lmap(self.vehicle.speed, self.config["reward_speed_range"], [0, 1])
+        reward = \
+            + self.COLLISION_REWARD * self.vehicle.crashed \
+            + self.LEFT_LANE_REWARD * lane / max(len(neighbours) - 1, 1) \
+            + self.HIGH_SPEED_REWARD * np.clip(scaled_speed, 0, 1)
+        reward = utils.lmap(reward,
+                          [self.COLLISION_REWARD, self.HIGH_SPEED_REWARD + self.LEFT_LANE_REWARD],
+                          [0, 1])
+        reward = 0 if not self.vehicle.on_road else reward
+        return reward
+
+    def _is_terminal(self) -> bool:
+        """
+        The episode is over if the ego vehicle crashed or the time is out.
+        """
+        return self.vehicle.crashed or \
+            self.steps >= self.config["duration"] or \
+            (self.config["offroad_terminal"] and not self.vehicle.on_road)
+
+    def _cost(self, action: int) -> float:
+        """
+        The constraint signal is the time spent driving on the opposite lane
+        and occurrence of collisions.
+        """
+        return float(self.vehicle.crashed) \
+               + float(self.vehicle.lane_index[2] == 0)/15
+
+    def _reset(self) -> np.ndarray:
+        self._make_road()
+        self._make_vehicles()
+
+    def _make_road(self, length=128):
+        """
+        Making double lane road with counter-clockwise U-Turn.
+        :return: the road
+        """
+        net = RoadNetwork()
+
+        # Defining lower starting lanes before the U-Turn.
+        net.add_lane("c", "d", StraightLane([length, 0], [0, 0],
```

Comment:
Aren't the lanes from/to labels mixed up ?
I think it should be:
* lower segment (before turn): A -> B
* turn: B -> C
* upper segment (after turn): C -> D

But it seems to me that you swapped upper and lower segments... Not sure why it still works though ^^

Reply:
Yes, you are right in your findings. I realized it now there is a mistake in my added comments. I'll update those comments as they are misleading.